components/br-stream/src/event_loader.rs

```diff
@@ -0,0 +1,235 @@
+                    result.push(write);
+                    if !default.0.is_empty() {
+                        let default = ApplyEvent::from_committed(
+                            CF_DEFAULT,
+                            default.0,
+                            default.1,
+                            self.region_id,
+                        )?;
+                        result.push(default);
+                    }
+                }
+            }
+        }
+        Ok(self.scanner.take_statistics())
+    }
+}
+
+/// The context for loading incremental data between range.
+/// Like [`cdc::Initializer`], but supports initialize over range.
+/// Note: maybe we can merge those two structures?
+#[derive(Clone)]
+pub struct InitialDataLoader<E, R, RT> {
+    router: RT,
+    regions: R,
+    // Note: maybe we can make it an abstract thing like `EventSink` with
+    //       method `async (KvEvent) -> Result<()>`?
+    sink: Router,
+    store_id: u64,
+
+    _engine: PhantomData<E>,
+}
+
+impl<E, R, RT> InitialDataLoader<E, R, RT>
+where
+    E: KvEngine,
+    R: RegionInfoProvider + Clone + 'static,
+    RT: RaftStoreRouter<E>,
+{
+    pub fn new(router: RT, regions: R, sink: Router, store_id: u64) -> Self {
+        Self {
+            router,
+            regions,
+            sink,
+            store_id,
+            _engine: PhantomData,
+        }
+    }
+
+    /// Start observe over some region.
+    /// This will register the region to the raftstore as observing,
+    /// and return the current snapshot of that region.
+    pub fn observe_over(&self, region: &Region, cmd: ChangeObserver) -> Result<impl Snapshot> {
+        // There are 2 ways for getting the initial snapshot of a region:
+        //   1. the BR method: use the interface in the RaftKv interface, read the key-values directly.
+        //   2. the CDC method: use the raftstore message `SignificantMsg::CaptureChange` to
+        //      register the region to CDC observer and get a snapshot at the same time.
+        // Registering the observer to the raftstore is necessary because we should only listen events from leader.
+        // In CDC, the change observer is per-delegate(i.e. per-region), we can create the command per-region here too.
+
+        let (callback, fut) = tikv_util::future::paired_future_callback();
+        self.router
+            .significant_send(
+                region.id,
+                SignificantMsg::CaptureChange {
+                    cmd,
+                    region_epoch: region.get_region_epoch().clone(),
+                    callback: Callback::Read(Box::new(|snapshot| callback(snapshot.snapshot))),
+                },
+            )
+            .map_err(|err| {
+                annotate!(
+                    err,
+                    "failed to register the observer to region {}",
+                    region.get_id()
+                )
+            })?;
+        let snap = block_on(fut)
```

Comment:
🤔 

`observe_over` cannot be `async`?

Reply:
Yes, because `KvEngine` isn't `Sync` hence we cannot make it a `Send` future.